### PR TITLE
Do not use double for single-prec upsample

### LINKED ISSUE #87968: [PT 1.13] resize bilinear yields different results in newer versions
### 🐛 Describe the bug

First, I tried to generate the inputs with the following code with PyTorch 1.13.
```python
>>> import torch
>>> import torch.nn.functional as F
>>> t = torch.randn(1, 3, 128, 128)
>>> torch.save(t, 'bilinear_inputs.pt')
>>> ot = F.interpolate(t, scale_factor=2.0, mode='bilinear', align_corners=True)
>>> torch.save(ot, 'bilinear_outputs_1_13.pt')
```

Then I tried to do the comparison with the Python environment with PyTorch 1.10.
```python
>>> import torch
>>> import torch.nn.functional as F
>>> t = torch.load('bilinear_inputs.pt')
>>> ot = torch.load('bilinear_outputs_1_13.pt')
>>> y = F.interpolate(t, scale_factor=2.0, mode='bilinear', align_corners=True)
>>> torch.allclose(ot, y)
False
>>> torch.testing.assert_close(ot, y)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.8/site-packages/torch/testing/_asserts.py", line 971, in assert_close
    raise error_meta.to_error()
AssertionError: Tensor-likes are not close!

Mismatched elements: 609 / 196608 (0.3%)
Greatest absolute difference: 2.562999725341797e-05 at index (0, 2, 131, 145) (up to 1e-05 allowed)
Greatest relative difference: 0.07112191124437972 at index (0, 1, 135, 191) (up to 1.3e-06 allowed)
```

As can be seen, the results are not within numerical differences caused by floating point representation.

Attachments:
[bilinear_data.zip](https://github.com/pytorch/pytorch/files/9886015/bilinear_data.zip)

### Versions

PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.6 (x86_64)
GCC version: Could not collect
Clang version: 12.0.0 (clang-1200.0.32.28)
CMake version: version 3.18.0
Libc version: N/A

Python version: 3.8.6 (default, Oct  8 2020, 14:06:32)  [Clang 12.0.0 (clang-1200.0.32.2)] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.4
[pip3] onnx2pytorch==0.4.1
[pip3] onnx2torch==1.3.0
[pip3] pytorch-lightning==1.6.3
[pip3] torch==1.10.0
[pip3] torchmetrics==0.8.2
[pip3] torchvision==0.11.1
[conda] Could not collect

cc @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by malfet**
I'm not sure, what would be the best behaviour here, but it feels a bit strange to perform parts of `float32` computations as `float64` and then downcast them back to `float32`.

Use `at::opmath_type` rather than `at:acc_type` as no accumulation is used in the op.

I don't know much about double vs single precision scalar perf on x86 CPU, but before the change:
```
python -c "import timeit;import torch;x=torch.arange(100, dtype=torch.float32).reshape(1, 1, 10, 10); print(timeit.Timer(stmt='torch.nn.functional.interpolate(x, scale_factor=2.0, mode=\"bilinear\", align_corners=False)', globals={'x':x, 'torch':torch}).timeit())"
11.337517574429512
```
After the change:
```
$ python -c "import timeit;import torch;x=torch.arange(100, dtype=torch.float32).reshape(1, 1, 10, 10); print(timeit.Timer(stmt='torch.nn.functional.interpolate(x, scale_factor=2.0, mode=\"bilinear\", align_corners=False)', globals={'x':x, 'torch':torch}).timeit())"
10.513805857859552
```
I.e. roughly 7% perf degradation (measured on Intel(R) Xeon(R) Platinum 8275CL CPU @ 3.00GHz) 

NOTE:
 - `aten::acc_type<float, false>` yields `double`
 - `aten::acc_type<float, true>` return `float`.

Fixes https://github.com/pytorch/pytorch/issues/87968

cc @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/UpSample.h [hunk 1]
```diff
@@ -447,7 +447,7 @@ static inline void compute_source_index_and_lambda(
     lambda0 = static_cast<scalar_t>(1);
     lambda1 = static_cast<scalar_t>(0);
   } else {
-    using accscalar_t = at::acc_type<scalar_t, false>;
+    using accscalar_t = at::acc_type<scalar_t, true>;
```
Comment (ngimel) [inline aten/src/ATen/native/UpSample.h:450]: nit: why don't we use `opmath_t` here that doesn't have weird discrepancy between use_cuda and not?

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp [hunk 2]
```diff
@@ -975,7 +975,7 @@ struct HelperInterpCubic : public HelperInterpBase {
 
         int64_t * idx_ptr;
         scalar_t * wt_ptr;
-        using accscalar_t = at::acc_type<scalar_t, false>;
+        using accscalar_t = at::acc_type<scalar_t, true>;
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/UpSampleKernel.cpp:978]: nit: do `opmath_t` here as well?

**Comment by jgong5** [comment]
@malfet May I know what result we are going to match here, 1.10/1.11 or 1.12? If you check the code from 1.12 (https://github.com/pytorch/pytorch/blob/release/1.12/aten/src/ATen/native/cpu/UpSampleKernel.cpp), there is already `acc_type` introduced in the code, even though in 1.13, the fix in https://github.com/pytorch/pytorch/pull/83847 tried to address the acc_type completely. If we revert all the usage of `acc_type` in the code, it would be a regression between 1.12 and 1.13? On the other hand, do you think it is an issue or an improvement in 1.13? The double precision accumulator improves accuracy as we know. There are also a lot of places where `acc_type<scalar_t, false>` is used for float type, e.g., https://github.com/pytorch/pytorch/blob/release/1.13/aten/src/ATen/native/cpu/batch_norm_kernel.cpp#L176. So it is more like an improvement than an issue?

**Review by mingfeima — APPROVED** [review]

**Comment by jgong5** [comment]
According to the [investigation](https://github.com/pytorch/pytorch/issues/87968#issuecomment-1297813602) from @vfdev-5 , there seems no diff between 1.10 and 1.12. I discussed offline with @mingfeima and am ok to revert it back with `opmath_t`. @malfet

**Comment by mingfeima** [comment]
@malfet the perf regression should not be a big issue, the benchmarked shape `[1, 1, 10, 10]` is probably too small to get a solid result.

Major performance bottleneck of  `interpolation` on `channels first memory format` is the `index` and `weight` calculation (a lot of data type conversions). To handle this, the `index` and `weight` are cached in the current code base:

1. **original version**: for each element of [n, c, h, w], need to calculate `index_h/weight_h` and `index_w/weight_w`, so the complexity is N * C* H * W * 2; on channels last memory format the `index/weight` calculation is not a big issue since All `C` would be sharing the same `index/weight`, so the complexity is `N * H * W * 2`
2. **the current code base**: since each row would have the same `index_h/weight_h` and each column would have the same `index_w/weight_w`, we can do pre-calculation, and the complexity is reduced to H * 2 + W * 2; with this type optimization, the calculation of `index` and `weight` is **NO longer** the major perf bottleneck.
3. **additional improvement**: the current code base is mostly non-vectorized logic (this operation is `indirect memory access`, e.g. `dst[i] = src[index[i]]` which will be mapped to `gather` on simd and GCC won't do auto-vectorization for it). #87053 is targeting at vectorizing this kernel, so as to make sure interpolation on PyTorch would not be slower than PIL.

**Review by ngimel — APPROVED** [review]
There's a stray accscalar_t, otherwise lgtm

**Review by jgong5 — APPROVED** [review]
LGTM. Please fix the acc_type line that I pointed out.

**Pull request closed without merging**